clang/include/clang/AST/PreorderAST.h

```diff
@@ -46,6 +46,64 @@ namespace clang {
           assert(!isa<LeafExprNode>(Parent) &&
                  "Parent node cannot be a LeafExprNode");
       }
+
+    // Recursively coalesce BinaryOperatorNodes having the same commutative
+    // and associative operator.
+    // @param[in] this is the current node of the AST.
+    // @param[in] Changed indicates whether a node was coalesced. We need this
+    // to control when to stop recursive coalescing.
+    // @param[in] Error indicates whether an error occurred during coalescing.
+    virtual void Coalesce(bool &Changed, bool &Error) { }
+
+    // Recursively descend a Node to sort the children of all
+    // BinaryOperatorNodes if the binary operator is commutative.
+    // @param[in] this is the current node of the AST.
+    // @param[in] Lex is used to lexicographically compare Exprs and Decls
+    // that occur within nodes.
+    virtual void Sort(Lexicographic Lex) { }
+
+    // Constant fold integer expressions.
+    // @param[in] this is the current node of the AST.
+    // @param[in] Changed indicates whether constant folding was done. We need
+    // this to control when to stop recursive constant folding.
+    // @param[in] Error indicates whether an error occurred during constant
+    // folding.
+    // @param[in] Ctx is used to create constant expressions.
+    virtual void ConstantFold(bool &Changed, bool &Error, ASTContext &Ctx) { }
+
+    // Compare nodes according to their kind.
+    // @param[in] this is the current node of the AST.
+    // @param[in] Other is the node to compare to this.
+    // @return Returns a Lexicographic::Result indicating the comparison
+    // between this and Other according to their node kinds.
+    Result CompareKinds(const Node *Other) const {
+      if (Kind < Other->Kind)
+        return Result::LessThan;
+      if (Kind > Other->Kind)
+        return Result::GreaterThan;
+      return Result::Equal;
+    }
+
+    // Compare two nodes lexicographically.
+    // @param[in] this is the current node of the AST.
+    // @param[in] Other the node to compare to this.
+    // @param[in] Lex is used to lexicographically compare Exprs and Decls
+    // that occur within nodes.
+    // @return Returns a Lexicographic::Result indicating the comparison
+    // between this and Other.
+    virtual Result Compare(const Node *Other, Lexicographic Lex) const {
+      return CompareKinds(Other);
+    }
+
+    // Print the node.
+    // @param[in] this is the current node of the AST.
+    virtual void PrettyPrint(llvm::raw_ostream &OS, ASTContext &Ctx) const { }
+
+    // Cleanup the memory consumed by this node.
+    // @param[in] this is the current node of the AST.
+    virtual void Cleanup() {
```

Comment:
The Node class must have a virtual destructor otherwise when you invoke delete on the derived class object the base class destructor won't be called. I also see a compile-time warning for this:

`PreorderAST.cpp:410:14: warning: deleting object of polymorphic class type ‘clang::Node’ which has non-virtual destructor might cause undefined behavior [-Wdelete-non-virtual-dtor]
  410 |     delete(*I);`